### ISSUE TITLE
telethon.errors.rpcerrorlist.PasswordHashInvalidError: The password (and thus its hash value) you entered is invalid

### ISSUE THREAD

**Issue by masoudcom** [LonamiWebs/Telethon#1029]
Hi.
When use `edit_2fa` for first time, get error `telethon.errors.rpcerrorlist.PasswordHashInvalidError`!
telethon version is : 1.3

**Comment by masoudcom** [comment]
verbose:
 File "<stdin>", line 1, in <module>
  File "/home/uroot/.local/lib/python3.6/site-packages/telethon/sync.py", line 35, in syncified
    else asyncio.get_event_loop().run_until_complete(coro)
  File "/usr/lib/python3.6/asyncio/base_events.py", line 468, in run_until_complete
    return future.result()
  File "/home/uroot/.local/lib/python3.6/site-packages/telethon/client/auth.py", line 470, in edit_2fa
    current_password_hash, new_settings=new_settings
  File "/home/uroot/.local/lib/python3.6/site-packages/telethon/client/users.py", line 60, in __call__
    result = await future
telethon.errors.rpcerrorlist.PasswordHashInvalidError: The password (and thus its hash value) you entered is invalid

**Comment by Lonami** [comment]
Aw, too bad there's people actually interested in the feature. That makes me have less reasons to remove the function altogether. Although compatibility was enough reason to keep it. Telegram has probably changed things server-side and, although the library is using the right method, it's calculating 
it incorrectly. You see:

https://github.com/LonamiWebs/Telethon/blob/4562ba9ccfb1aef33facd5ca1a2675705d390e1c/telethon/client/auth.py#L474-L485

And if we look at the `schema.tl` for the current layer:

https://github.com/LonamiWebs/Telethon/blob/4562ba9ccfb1aef33facd5ca1a2675705d390e1c/telethon_generator/data/telegram_api.tl#L947

https://github.com/LonamiWebs/Telethon/blob/4562ba9ccfb1aef33facd5ca1a2675705d390e1c/telethon_generator/data/telegram_api.tl#L1185

So currently the library is stuck at layer 82 because [layer 86 has added some new stuff](https://github.com/telegramdesktop/tdesktop/blob/19e407532b047702a78d335473f191484e105f58/Telegram/Resources/scheme.tl#L1027-L1037) regarding passwords… And I don't have enough interest in making it work myself. Technically, it's possible to read on [how `tdesktop` does it](https://github.com/telegramdesktop/tdesktop/blob/18b74b90451a7db2379a9d753c9cbaf8734b4d5d/Telegram/SourceFiles/core/core_cloud_password.cpp), and eventually we should do that.

**TL; DR;** would be nice if someone could invest some of their time into porting the new way of working with passwords to Telethon, from `tdesktop` or any other implementation.

**Comment by masoudcom** [comment]
client.edit_2fa(new_password='mser', hint='Reg')

**Comment by Kyle2142** [comment]
Thank you for the reply, however:
![image](https://user-images.githubusercontent.com/19349418/47394000-8f210a80-d721-11e8-869c-554cd75f1334.png)
Can not reproduce...
Perhaps the method only works for people who have used it prior to the new layer?
Side note: I am the one who made (well, ported) the original function in telethon and I apologize for not having the energy to implement the new one.

**Comment by Lonami** [comment]
Closed since the new `edit_2fa` exists and has been tested.

**Comment by timoilya** [comment]
I have same exception and i know how reproduce it:

```
from telethon.sync import TelegramClient
from telethon.sessions.string import StringSession

string_session = "<session_string>"

password = 'secret_password'

with TelegramClient(StringSession(string_session), api_id=123, api_hash='123') as client:
    client.edit_2fa(new_password=password)

```

`string_session` value I will send in private message. reproducing the error depends on `string_session`.

this problem is reproduced on the recent version 1.21.1
